content-based-message-routing/main.bal

```diff
@@ -0,0 +1,85 @@
+import ballerina/http;
+import ballerina/log;
+
+type Patient record {|
+    string name;
+    string dob;
+    string ssn;
+    string address;
+    string phone;
+    string email;
+|};
+
+type ReservationRequest record {|
+    Patient patient;
+    string doctor;
+    string hospital_id?;
+    string hospital;
+    string appointment_date;
+|};
+
+type Doctor record {|
+    string name;
+    string hospital;
+    string category;
+    string availability;
+    float fee;
+|};
+
+type ReservationResponse record {|
+    int appointmentNumber;
+    Doctor doctor;
+    Patient patient;
+    float fee;
+    string hospital;
+    boolean confirmed;
+    string appointmentDate;
+|};
+
+configurable int port = 8290;
+
+final http:Client grandOaksClient = check new ("http://localhost:9090/grandoaks/categories");
+final http:Client clemencyClient = check new ("http://localhost:9090/clemency/categories");
+final http:Client pineValleyClient = check new ("http://localhost:9090/pinevalley/categories");
+
+service /healthcare on new http:Listener(8290) {
+    resource function post categories/[string category]/reserve(
+            @http:Payload ReservationRequest reservationRequest
+        ) returns ReservationResponse|http:NotFound|http:InternalServerError? {
+
+        string hospital = reservationRequest.hospital;
+        ReservationResponse|http:ClientError? resp = ();
+        match hospital {
+            "grand oak community hospital" => {
+                log:printInfo("Routed to grand oak community hospital");
+                resp = grandOaksClient->/[category]/reserve.post(reservationRequest);
+            }
+            "clemency medical center" => {
+                log:printInfo("Routed to clemency medical center");
+                resp = clemencyClient->/[category]/reserve.post(reservationRequest);
+            }
+            "pine valley community hospital" => {
+                log:printInfo("Routed to pine valley community hospital");
+                resp = pineValleyClient->/[category]/reserve.post(reservationRequest);
+            }
+            _ => {
+                log:printError(string `Routed to none. Hospital not found: ${hospital}`);
+                return <http:NotFound>{body: string `Hospital not found: ${hospital}`};
+            }
+        }
+
+        if resp is ReservationResponse {
+            log:printInfo("Reservation request confirmed");
+            return resp;
+        }
+
+        if resp is http:ClientRequestError && resp.detail().statusCode == 404 {
+            log:printError("Reservation request failed", resp);
+            string cuase = resp.detail().body.toString();
+            return <http:NotFound>{body: cuase};
```

Comment:
Typo. Also why do we need to include the entire detail?

Reply:
Because I want to send the error messages coming from the services to the client. Here, we get error message in the body as a string

Reply:
Isn't the error message enough? Sending the entire detail may not be ideal (in the general case), instead we would extract out parts we need.

Reply:
We can give a basic error out IMO. Please check the 1st tutorial too.

Reply:
Ack. Thanks